crates/wast/src/core/resolve/names.rs

```diff
@@ -46,16 +47,41 @@ impl<'a> Resolver<'a> {
     }
 
     fn register_type(&mut self, ty: &Type<'a>) -> Result<(), Error> {
+        let type_index = self.types.register(ty.id, "type")?;
+
         match &ty.def {
             // For GC structure types we need to be sure to populate the
             // field namespace here as well.
             //
-            // The field namespace is global, but the resolved indices
-            // are relative to the struct they are defined in
+            // The field namespace is relative to the struct fields are defined in
             TypeDef::Struct(r#struct) => {
                 for (i, field) in r#struct.fields.iter().enumerate() {
                     if let Some(id) = field.id {
-                        self.fields.register_specific(id, i as u32, "field")?;
+                        if let Some(parent_type_index) = ty
+                            .parent
+                            .map(|index| self.types.get_index(&index))
+                            .flatten()
+                        {
+                            if let Some(parent_field_ns) = self.fields.get(&parent_type_index) {
+                                let field_idx =
+                                    parent_field_ns.resolve(&mut Index::Id(id), "field")?;
+                                if field_idx != i as u32 {
```

Comment:
Is this strictly necessary or something that could be deferred to type-checking? Ideally the text format is the bare minimum to have developer conveniences like `$foo` but also not enough to duplicate the validation that happens on the binary format. This check here I think may be ok to remove and rely on the binary format validation to perform?

Reply:
> Is this strictly necessary or something that could be deferred to type-checking?

I think it is: I didn't see if names were preserved in the binary format (pretty sure they are not). I definitely would want to defer it if they were.

And this check is about names only: a field with the same name should be at the same index as in the parent type. None of the binary format validation is happening here.

This seems appropriate: names are only available in text (to my knowledge), so we check them when parsing text.

Reply:
Removed as a result of discussion below: https://github.com/bytecodealliance/wasm-tools/pull/1117#discussion_r1260321905